dev/doc/parsing.md

```diff
@@ -0,0 +1,378 @@
+  added productions in specific ways to ensure that their additions are parsable and that they
+  don't break existing productions.
+
+## Contents ##
+
+- [Grammars: `*.mlg` File Structure](#grammars-mlg-file-structure)
+- [Grammars: Nonterminals and Productions](#grammars-nonterminals-and-productions)
+  - [Alternate production syntax](#alternate production-syntax)
+- [Usage notes](#usage-notes)
+  - [Other components](#other-components)
+  - [Parsing productions](#parsing-productions)
+  - [Lookahead](#lookahead)
+
+## Grammars: `*.mlg` File Structure ##
+
+Grammars are defined in `*.mlg` files, which `coqpp` compiles into `*.ml` files at build time.
+`coqpp` code is in `coqpp` directory.  `coqpp` uses yacc and lex to parse the grammar files.
+You can examine its yacc and lex input files in `coqpp_lex.mll` and `coqpp_parse.mly` for
+details not fully covered here.
+
+In addition, there is a `doc_grammar` build utility that uses the `coqpp` parser to extract the
+grammar, then edits and inserts it into the documentation.  This is described in `doc/tools/docgram/README.md`.
+`doc_grammar` generates `doc/tools/docgram/fullGrammar`, which has the full grammar for Coq
+(not including some optionally-loaded plugins).  This may be easier to read since everything is
+in one file and the parser action routines and other OCaml code are omitted.
+
+`*.mlg` files contain the following types of nodes (See `node` in the yacc grammar).  This part is
+very specific to Coq (not so similar to Camlp5):
+
+* OCaml code - OCaml code enclosed in curly braces, which is copied verbatim to the generated `*.ml` file
+
+* Comments - comments in the `*.mlg` file in the form `(* ... *)`, which are dropped.  Comments
+  in OCaml code are preserved.
+
+* `DECLARE_PLUGIN "*_plugin"` - associates the file with a specific plugin, for example "ltac_plugin"
+
+* `GRAMMAR EXTEND` - adds additional nonterminals and productions to the grammar and declares global
+  nonterminals referenced in the `GRAMMAR EXTEND`:
+
+  ```
+  GRAMMAR EXTEND Gram
+    GLOBAL:
+      bignat bigint ...;
+    <nonterminal definitions>
+  END
+  ```
+
+  TODO: What is `Gram` above?  Worth explaining?
+
+  Global nonterminals are declared in `pcoq.ml`, e.g. `let bignat = Entry.create "Prim.bignat"`.
+  All the `*.mlg` files include `open Pcoq` and often its modules, e.g. `open Pcoq.Prim`.
+
+* `VERNAC COMMAND EXTEND` - adds new command syntax by adding productions to the
+  `command` nonterminal.  For example:
+
+  ```
+  VERNAC COMMAND EXTEND ExtractionLibrary CLASSIFIED AS QUERY
+  | [ "Extraction" "Library" ident(m) ]
+    -> { extraction_library false m }
+  END
+  ```
+
+  TODO: explain "ExtractionLibrary", CLASSIFIED AS, CLASSIFIED BY, "{ tactic_mode }", STATE
+
+  Productions here are represented with alternate syntax, described later.
+
+  It's also possible to add productions to `command` with `GRAMMAR EXTEND` but
+  without the additional fields.
+
+  TODO: Which is the best practice?
```

Comment:
Just extending the grammar with `GRAMMAR EXTEND` will fail at runtime, since you need to register the classifiers and whatnot. You shouldn't do that.